analyzer/evmverifier/sourcify/client.go

```diff
@@ -0,0 +1,210 @@
+//
+// API docs: https://docs.sourcify.dev/docs/api/repository/get-file-static/
+//
+// Note: this could also use the repository API (via IPFS proxy) instead:
+// - https://docs.sourcify.dev/docs/api/repository/get-file-repository/
+func (s *SourcifyClient) GetContractMetadata(ctx context.Context, runtime common.Runtime, address ethCommon.Address) (*Metadata, error) {
+	// Fetch contract metadata.json.
+	u := *s.serverUrl
+	u.Path = path.Join(u.Path, "repository/contracts/full_match", sourcifyChains[s.chain][runtime], address.String(), "metadata.json")
+	body, err := s.callAPI(ctx, http.MethodGet, &u)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch contract metadata.json: %w", err)
+	}
+
+	// Parse response.
+	var response Metadata
+	if err := json.Unmarshal(body, &response); err != nil {
+		return nil, fmt.Errorf("failed to parse contract metadata.json: %w", err)
+	}
+
+	return &response, nil
+}
+
+// GetContractSourceFiles returns the source files for the given contract address that is verified on sourcify.
+//
+// API docs: https://docs.sourcify.dev/docs/api/server/get-source-files-full/
+//
+// Note: this could also use the repository API (via IPFS proxy) instead:
+// - https://docs.sourcify.dev/docs/api/repository/get-file-repository/
+func (s *SourcifyClient) GetContractSourceFiles(ctx context.Context, runtime common.Runtime, address ethCommon.Address) ([]map[string]*json.RawMessage, *Metadata, error) {
+	// Fetch contract source files.
+	u := *s.serverUrl
+	u.Path = path.Join(u.Path, "files", sourcifyChains[s.chain][runtime], address.String())
+	body, err := s.callAPI(ctx, http.MethodGet, &u)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to fetch contract source files: %w", err)
+	}
+
+	// Ensure response is a valid JSON.
+	var response []map[string]*json.RawMessage // Response is an array of source files.
+	if err = json.Unmarshal(body, &response); err != nil {
+		return nil, nil, fmt.Errorf("failed to parse contract source files: %w", err)
+	}
+
+	// Parse metadata.json.
+	var metadata Metadata
+	var found bool
+	for _, file := range response {
+		var name string
+		if err = json.Unmarshal(*file["name"], &name); err != nil {
+			continue
+		}
+		if name == "metadata.json" {
+			// Remove the escaped quotes from the metadata json content string.
+			content, err := strconv.Unquote(string(*file["content"]))
+			if err != nil {
+				s.logger.Error("failed to unquote contract metadata string", "err", err, "content", content)
+				return nil, nil, err
+			}
+			if err = json.Unmarshal([]byte(content), &metadata); err != nil {
+				s.logger.Error("failed to unmarshal contract metadata", "err", err, "content", content)
+				return nil, nil, fmt.Errorf("failed to unmarshal contract metadata: %w", err)
+			}
+			found = true
+			break
+		}
+	}
+	if !found {
+		return nil, nil, fmt.Errorf("failed to find metadata.json in source files")
```

Comment:
Not sure. I think that realistically, the metadata.json should always be there, but there's no assurance from the API that it will be there.

That said then it doesn't really matter so changing this to a warning is fine with me.


Just got an idea. I can check the ~50k verified contracts on the ethereum sourcify API and see if metadata.json is always found.

Reply:
Let's change this to a warning. I actually don't expect it will ever matter either. But if there ever is a contract with no metadata.json, a warning will still download the files and ignore the contract from there on; whereas a `return nil, nil, Error()` will make the analyzer reattempt the download every single cycle, waste our quota and pollute the logs.

Reply:
Updated these to warnings as suggested. 

I'm also in the process of running a test on all 50k Ethereum Sourcify contracts and checking if metadata.json can be found in all cases (and ABI successfully parsed as well).

Reply:
fwiw: tested on ~30k ethereum contracts over night. No errors in finding the contract metadata json file.